### ISSUE TITLE
[Feature Request] - Implement kittys terminal graphics protocol

### ISSUE THREAD

**Issue by misaelaguayo** [zellij-org/zellij#2814]
It would be neat if zellij could use kitty terminal graphics to display images or other files. Tmux had a [similar](https://github.com/tmux/tmux/issues/1391) issue closed, it seems like they will not be implementing it. Zellij does provide support for sixels but this is a very old protocol with not much support from other applications.

**Comment by daveman1010221** [comment]
Is there a sense of the work required to implement this?

**Comment by B4mbus** [comment]
> FIY: On the yazi wiki, there's a fix to use it with tmux: https://github.com/sxyazi/yazi/wiki/Image-preview-within-tmux This looks like a workaround, not sure how it works exactly but maybe zellij can do something similar?

Yes, but it's extremely slow on tmux for some reason.

**Comment by B4mbus** [comment]
> Is there a sense of the work required to implement this?

There's a plethora of users that want to enjoy images in their terminal and some terminal tools implement image support (image.nvim plugin for neovim, neorg plugin for neovim, a lot of file browsers, etc.). I almost never leave terminal and currently use tmux, image implementation in zellij would be a dealbreaker for me.

**Comment by sxyazi** [comment]
I really want to see Zellij support kitty graphics protocol, which is currently the most complete and advanced.

It supports Unicode placeholders, I think this could solve the stuttering issue in Zellij's Sixel implementation. Kitty protocol simply prints some escape sequence to the screen, and terminals that support that protocol display it as an image and accurately render the image position when scrolling.

This process is transparent to Zellij, and it doesn't need to parse and understand the meaning of the protocol itself, making it efficient.

I've just implemented Unicode placeholders for Yazi and tested it in kitty+tmux, it works well:

https://github.com/zellij-org/zellij/assets/17523360/7a40ed76-e406-4615-8029-6d3c7d636926

**Comment by musjj** [comment]
@sxyazi 
Is there something that needs to be done on the zellij side for this to work? I'm testing this on yazi's latest commit, but image previews still doesn't work.

**Comment by sxyazi** [comment]
> Is there something that needs to be done on zellij's side for this to work? I'm testing yazi (with the latest commit) from inside zellij, but image previews still doesn't work.

Yes, Zellij does not support it; this requires Zellij to provide support.

> I've also tested it under tmux and it kinda works, but everything is really buggy (overlapping images, preview not updating, etc.) for some reason.

What are your Yazi version and tmux version? What terminal are you using? Please also do an `echo $TERM, $TERM_PROGRAM, $XDG_SESSION_TYPE` and paste the output here.

**Comment by AnonymouX47** [comment]
Just came across this and thought I should chime in.

I believe what's required on zellij's end is a Passthrough sequence (I don't know if such exists yet but I suspect not) i.e some something like [this](https://github.com/tmux/tmux/wiki/FAQ#what-is-the-passthrough-escape-sequence-and-how-do-i-use-it). No special handling of graphics control sequences should be required.

This should be all that's required to get Kitty's virtual placements working, assuming Zellij already supports indexed- (and optionally, direct-) -color SGR sequences and passes them through as recieved.

It'll also allow the use of other image/graphics protocols **to a certain extent**, particularly in full-screen (non-scrolling) applications. For full/proper support, specialized implementations/handling will be required but personally, I'd discourage that for a terminal multiplexer.

**Comment by sweetbbak** [comment]
This is also how this issue was solved with junegunn/fzf [junegunn/fzf kitty image protocol passthrough fix](https://github.com/junegunn/fzf/commit/d8188fce7b7bea982e7f9050c35e488e49fb8fd0)

He uses a regex to detect the terminal sequence for images then assigns a no-op function for the preview that allows the un-altered terminal sequence to passthrough to the terminal so that the image is rendered like normal.

**Comment by hfytr** [comment]
is this still being worked on?

**Comment by misaelaguayo** [comment]
I don’t think anyone has started to implement this. Just been talking about ways in which it could be achieved

**Comment by misaelaguayo** [comment]
I've been digging into this a bit more since I would like to see it implemented. It seems that zellij uses [alacritty's vte parser](https://github.com/alacritty/vte) to interpret ANSI/VT codes in `terminal_pane.rs`. Perhaps in the `handle_pty_bytes` function we can do a similar approach to what was done for fzf and do a no op and not pass in those bytes to the `vte_parser`

**Comment by quadespresso** [comment]
Fwiw I've tried the following:

- Kitty + yazi :heavy_check_mark: 
- Kitty + tmux + yazi :heavy_check_mark: 
- Kitty + Zellij + yazi :x: 
- Wezterm + yazi :heavy_check_mark: 
- Wezterm + tmux + yazi :heavy_check_mark: 
- Wezterm + Zellij + yazi :heavy_check_mark: 

```bash
❯ neofetch
██████████████████  ████████   jimc@kapua 
██████████████████  ████████   ---------- 
██████████████████  ████████   OS: Manjaro Linux x86_64 
██████████████████  ████████   Host: NH5x_NH7xHP 
████████            ████████   Kernel: 6.1.99-1-MANJARO 
████████  ████████  ████████   Uptime: 10 days, 15 hours, 22 mins 
████████  ████████  ████████   Packages: 3373 (pacman), 75 (nix-user), 48 (nix-default) 
████████  ████████  ████████   Shell: zsh 5.9 
████████  ████████  ████████   Resolution: 1920x1080 
████████  ████████  ████████   DE: GNOME 46.3.1 
████████  ████████  ████████   WM: Mutter 
████████  ████████  ████████   WM Theme: Custom-Accent-Colors 
████████  ████████  ████████   Theme: Adw-dark [GTK2/3] 
████████  ████████  ████████   Icons: Papirus-Dark-Maia [GTK2/3] 
                               Terminal: kitty 
                               CPU: 11th Gen Intel i7-11800H (16) @ 4.600GHz 
                               GPU: Intel TigerLake-H GT1 [UHD Graphics] 
                               GPU: NVIDIA GeForce RTX 3060 Mobile / Max-Q 
                               Memory: 41611MiB / 64087MiB 
```

**Comment by merikan** [comment]
@quadespresso That's great news, I didn't know it worked in Wezterm. Thank you for looking into this 👍

I've been using Alacritty for years and really missed the support for displaying images in the terminal and also ligatures. A few weeks ago decided to switch to Kitty and was somewhat surprised that images were not rendered in Zellij. After reading your comment I switched to Wezterm and everything works perfectly. 🥰 

A funny thing was that when I, in Wezterm, attached to a running session started with Kitty, the images were not displayed. I had to close the session and restart it.

```
$ fastfetch

OS: macOS Sonoma 14.5 arm64                                              
Host: MacBook Pro (16-inch, 2023)                                        
Kernel: Darwin 23.5.0                                                    
Uptime: 22 hours, 26 mins                                                
Packages: 42 (nix-default), 412 (brew), 133 (brew-cask)                  
Shell: zsh 5.9                                                           
Display (DELL U3419W): 3440x1440 @ 60 Hz in 34″ [External] *             
Display (Color LCD): 3456x2234 @ 120 Hz (as 1728x1117) in 16″ [Built-in] 
DE: Aqua                                                                 
WM: Quartz Compositor                                                    
WM Theme: Multicolor (Dark)                                              
Font: .AppleSystemUIFont [System], Helvetica [User]                      
Cursor: Fill - Black, Outline - White (32px)                             
Terminal: zellij 0.40.1                                                  
CPU: Apple M2 Max (12) @ 3,50 GHz                                        
GPU: Apple M2 Max (30) @ 1,40 GHz [Integrated]                           
Memory: 33,45 GiB / 64,00 GiB (52%)                                      
Swap: Disabled                                                           
Disk (/): 649,98 GiB / 926,35 GiB (70%) - apfs [Read-only]                                                   
Battery: 100% [AC connected]                                             
Power Adapter: 90W                                                       
Locale: en_US.UTF-8                                                      
```

**Comment by quadespresso** [comment]
@merikan interesting test scenario.

Based on that, I tried the opposite: I ran wezterm -> zellij -> yazi then navigated to an image to view.

Then I detached from the zellij session, reconnecting from a kitty terminal. It was a bit of a mess, an entire screenful of characters. Nothing vaguely resembling yazi was to be seen.

**Comment by merikan** [comment]
@quadespresso 
That was interesting. I had to try it too and can confirm that it just gets a bunch of characters all over the screen. A little surprising that it is not only the image on the right side that gets distorted, but also the directory structures etc.

I couldn't resist, but also had Alacritty connected to the same session and there everything looked ok, except that the image is not displayed as we already know.

Kind of fun to look at multiple clients connected to the same session when jumping around in Yazi 😁

**Comment by quadespresso** [comment]
@merikan it's always interesting to cross-check with alternatives. I ended up trying the same combo with Alacritty and saw the same, ie, yazi doesn't display the image, but is otherwise intact.

Had a quick go with a few other terminal emulators (Terminator, Terminology, GNOME Terminal, Tilix), but only got as far as doing a quick check with yazi (minus zellij) to verify it couldn't display images before proceeding beyond that. I couldn't recall whether any of those supported inline images, and figured this would be quicker than pulling up docs for each.

**Comment by hongquan** [comment]
@quadespresso Which version of "Wezterm + Zellij + yazi" that can display image?
My version doesn't work:
- WezTerm 20240203-110809-5046fc22
- Zellij 0.40.1
- Yazi 0.3.1

**Comment by sweetbbak** [comment]
None. Zellij would have to implement some form of a passthrough ansi sequence that wraps image protocol sequences that indicate that the sequence should be let through the multiplexer. [tmux/wiki/FAQ#what-is-the-passthrough-escape-sequence-and-how-do-i-use-it](https://github.com/tmux/tmux/wiki/FAQ#what-is-the-passthrough-escape-sequence-and-how-do-i-use-it) Im pretty sure zellij just strips the escape code which leaves a mess of base64 image data

currently `icat` for the Image kitty protocol automatically detects if its inside of tmux and uses unicode placeholder data and the special tmux ansi escape to wrap its encoded image data.

**Comment by merikan** [comment]
@quadespresso 
I have the same inconsistent behavior. I have two long running zellij sessions and when I open a new pane and run yazi, only one of them render the image. 🤔.  If I start a new session everything works ok.
I am trying to understand what is different between the two sessions but they are both started at about the same time and have version 0.40.1. 
Killing and restarting the "non-working" session, in the same wezterm window, did not help. If I kill and restart the "working" session, in the same wezterm window,  it still works.
If I kill and restart the the "non-working" session in a new wezterm window it starts rendering images. 🥳
```
zellij 0.40.1
wezterm 20240203-110809-5046fc22
Yazi 0.3.2 (Homebrew 2024-08-28)

```

**Comment by sxyazi** [comment]
I found this in the Zellij Discord server might be helpful (Quote from @imsnif in https://discord.com/channels/771367133715628073/771367133715628076/1272810902458204191):

> Hey - Zellij supports Sixel for displaying images in the terminal (assuming your terminal emulator supports it too ofc). There are some issues with Sixel that I hope will be solved in the next release, until then they can usually be fixed by clearing the Zellij cache (`~/.cache/zellij` by default on linux) and restarting.

**Comment by bentenjamin** [comment]
A side effect of this is nushell is unusable in zellij when its `use_kitty_protocol` config option is `true`, because whatever escape sequence is sent causes the output to be cleared, meaning you get zero output from any command

**Comment by bew** [comment]
> A side effect of this is nushell is unusable in zellij when its `use_kitty_protocol` config option is `true`, because whatever escape sequence is sent causes the output to be cleared, meaning you get zero output from any command

AFAIK the kitty graphics and keyboard protocol are separate and don't require each other, are you sure your comment is on the right issue?

**Comment by Straffern** [comment]
Is this an issue of either move to Wezterm or change to Tmux?
I as a Kitty user ran into this issue, when I tried to configure Image.nvim, and realizing other users had same problem with Zellij.

**Comment by merikan** [comment]
I can confirm that images does **not** display when using [Ghostty Terminal](https://ghostty.org/). Ghostty does support the Kitty graphics protocol and images are displayed in Yazi when not in Zellij. 

```
MacOS Sonoma 14.6.1 arm64
Zellij 0.41.2
Ghostty 1.0.0
Yazi 0.4.2
```

**Comment by vishalpaudel** [comment]
Ghostty+yazi works, Ghostty+zellij+yazi doesn't work.

Ghostty 1.0.1-arch
Yazi 0.4.2 (Arch Linux 2025-01-12)
zellij 0.41.2

**Comment by uroybd** [comment]
Please implement this.

**Comment by typkrft** [comment]
> Ghostty+yazi works, Ghostty+zellij+yazi doesn't work.
> 
> Ghostty 1.0.1-arch Yazi 0.4.2 (Arch Linux 2025-01-12) zellij 0.41.2

For what it's worth, for anyone looking into ghostty, the dev has said they will probably never implement sixel and critiqued libsixel pretty harshly. They currently use kitty image protocol.

**Comment by j4james** [comment]
The ghostty maintainer also said "This decision isn't final", and "I welcome any further discussion and feedback about it", so I don't think you can conclude they'll _never_ support Sixel. I'm also not sure why he brought up libsixel, because that's not something you'd use to implement Sixel in a terminal, but I'm guessing he didn't spend much time researching the topic, and there was a lot of misinformation being spread in that thread.

**Comment by luccahuguet** [comment]
> Ghostty+yazi works, Ghostty+zellij+yazi doesn't work.
> 
> Ghostty 1.0.1-arch Yazi 0.4.2 (Arch Linux 2025-01-12) zellij 0.41.2

going back to wezterm for now, thanks for the info people

**Comment by imroc** [comment]
Looking forward to this feature being implemented, wezterm+zellij+neovim+snacks.nvim cannot preview images

**Comment by imroc** [comment]
> kitty+zellij+neovim+snacks.nvim
> 
> I can't seem to get this working even when using Kitty. I usually zellij setup to open a session when I open a terminal window so that I am always inside zellij, but even if I use kitty as terminal, the image still will not display at all so I'm definitely looking forward to zellij having this feature

Any terminal will not work if you use zellij, because zellij does not support it. I use wezterm which also supports kitty graphics protocol and has enabled it, but it still does not work.

If zellij can support this configuration like tmux, it should be possible:

```bash
set -gq allow-passthrough on
```

**Comment by imroc** [comment]
> > > kitty+zellij+neovim+snacks.nvim
> > > I can't seem to get this working even when using Kitty. I usually zellij setup to open a session when I open a terminal window so that I am always inside zellij, but even if I use kitty as terminal, the image still will not display at all so I'm definitely looking forward to zellij having this feature
> > 
> > 
> > Any terminal will not work if you use zellij, because zellij does not support it. I use wezterm which also supports kitty graphics protocol and has enabled it, but it still does not work.
> > If zellij can support this configuration like tmux, it should be possible:
> > set -gq allow-passthrough on
> 
> I definitely had it work with wezterm before, so that's not entirely true. Though it might have been through a different graphics protocol - I am not sure about that.

Then it must be a different graphics protocol, like `yazi` support many image protocols, so it works with zellij, but `snacks.nvim` rely on kitty graphics protocol.

**Comment by j4james** [comment]
@imroc Using a passthrough mode like tmux would only work reliably with kitty unicode placeholders, which very few terminals support (I think only Kitty and Ghostty). So even if Zellij added passthrough, you still likely wouldn't be able to use `snacks.nvim` in Wezterm (at least until it adds placeholder support).

Frankly it would be a lot easier to just add sixel support to `snacks.nvim`. Inline images might not be feasible, but it already has a fallback mode without inline images which it uses on terminals that can't handle kitty placeholders. I'm sure it could just as easily do the same thing for sixel terminals.

**Comment by on3iro** [comment]
> > > > kitty+zellij+neovim+snacks.nvim
> > > > I can't seem to get this working even when using Kitty. I usually zellij setup to open a session when I open a terminal window so that I am always inside zellij, but even if I use kitty as terminal, the image still will not display at all so I'm definitely looking forward to zellij having this feature
> > > 
> > > 
> > > Any terminal will not work if you use zellij, because zellij does not support it. I use wezterm which also supports kitty graphics protocol and has enabled it, but it still does not work.
> > > If zellij can support this configuration like tmux, it should be possible:
> > > set -gq allow-passthrough on
> > 
> > 
> > I definitely had it work with wezterm before, so that's not entirely true. Though it might have been through a different graphics protocol - I am not sure about that.
> 
> Then it must be a different graphics protocol, like `yazi` support many image protocols, so it works with zellij, but `snacks.nvim` rely on kitty graphics protocol.

oh, sorry - I missed that you referred to snacks.nvim specifically

**Comment by c02y** [comment]
> Fwiw I've tried the following:
> 
>     * Kitty + yazi ✔️
> 
>     * Kitty + tmux + yazi ✔️
> 
>     * Kitty + Zellij + yazi ❌
> 
>     * Wezterm + yazi ✔️
> 
>     * Wezterm + tmux + yazi ✔️
> 
>     * Wezterm + Zellij + yazi ✔️
> 
> 
> ❯ neofetch
> ██████████████████  ████████   jimc@kapua 
> ██████████████████  ████████   ---------- 
> ██████████████████  ████████   OS: Manjaro Linux x86_64 
> ██████████████████  ████████   Host: NH5x_NH7xHP 
> ████████            ████████   Kernel: 6.1.99-1-MANJARO 
> ████████  ████████  ████████   Uptime: 10 days, 15 hours, 22 mins 
> ████████  ████████  ████████   Packages: 3373 (pacman), 75 (nix-user), 48 (nix-default) 
> ████████  ████████  ████████   Shell: zsh 5.9 
> ████████  ████████  ████████   Resolution: 1920x1080 
> ████████  ████████  ████████   DE: GNOME 46.3.1 
> ████████  ████████  ████████   WM: Mutter 
> ████████  ████████  ████████   WM Theme: Custom-Accent-Colors 
> ████████  ████████  ████████   Theme: Adw-dark [GTK2/3] 
> ████████  ████████  ████████   Icons: Papirus-Dark-Maia [GTK2/3] 
>                                Terminal: kitty 
>                                CPU: 11th Gen Intel i7-11800H (16) @ 4.600GHz 
>                                GPU: Intel TigerLake-H GT1 [UHD Graphics] 
>                                GPU: NVIDIA GeForce RTX 3060 Mobile / Max-Q 
>                                Memory: 41611MiB / 64087MiB

If anyone is interested and using Wayland, I'm using Hyprland, install [Überzug++](https://github.com/jstkdng/ueberzugpp) package, then `kitty+zellij+yazi` image/video preview will work, but the file is previewed in the top-middle of the screen instead of normally in the preview column. We can treat it as a workaround.

**Comment by misaelaguayo** [comment]
<img width="1711" alt="Image" src="https://github.com/user-attachments/assets/cf49e030-15a0-4cd6-b086-c97d9c54e810" />

I've been looking into this a bit more and implemented a crude passthrough. Kitty uses APC control characters for their images. This is ignored by zellij, so we should be able to safely pass this through to the terminal. 

The relevant changes were made in `handle_pty_bytes` in `terminal_pane` and `grid.rs`

<img width="541" alt="Image" src="https://github.com/user-attachments/assets/4db42741-0565-4a24-801e-fff37108a4a9" />

**Comment by imroc** [comment]
> <img alt="Image" width="1711" src="https://private-user-images.githubusercontent.com/29875928/457168041-cf49e030-15a0-4cd6-b086-c97d9c54e810.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTAzODQzMDcsIm5iZiI6MTc1MDM4NDAwNywicGF0aCI6Ii8yOTg3NTkyOC80NTcxNjgwNDEtY2Y0OWUwMzAtMTVhMC00Y2Q2LWIwODYtYzk3ZDljNTRlODEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjIwVDAxNDY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2Mzg1MWRhZmM2YTRmYTNjYmI0MjljNzY5NGQ3NGM4NjkyZjY0ZTcxMTU0NzU1Y2VjZmYxNjc0MjE4YmMyZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xZCe_cDwRQIuRX4nDfQY9p34UJoHbVWlUujrrHOEjQc">
> I've been looking into this a bit more and implemented a crude passthrough. Kitty uses APC control characters for their images. This is ignored by zellij, so we should be able to safely pass this through to the terminal.
> 
> The relevant changes were made in `handle_pty_bytes` in `terminal_pane` and `grid.rs`
> 
> <img alt="Image" width="541" src="https://private-user-images.githubusercontent.com/29875928/457168469-4db42741-0565-4a24-801e-fff37108a4a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTAzODQzMDcsIm5iZiI6MTc1MDM4NDAwNywicGF0aCI6Ii8yOTg3NTkyOC80NTcxNjg0NjktNGRiNDI3NDEtMDU2NS00YTI0LTgwMWUtZmZmMzcxMDhhNGE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjIwVDAxNDY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkMmUyN2JjZGRjNzIyNTUxOGI4NDljY2FmOTM3YjBhNDk4NWUxNGFkNWM0YjZiNmM0ZmI4ZTdjNjJiMGMxOGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.X8eytlLr3A0VdXGY4vObawxbz2ePK8-tBUxD-Kdv8wQ">

Please raise a PR, I really need it for my daily work

**Comment by misaelaguayo** [comment]
Here are the changes i've made if anyone is curious https://github.com/misaelaguayo/zellij/pull/1

**Comment by Andrew15-5** [comment]
It does show images now in the top left corner without clearing in `lf` + `ctpv`. For MP4, MP3, WebP files it gives `[PARSE ERROR] Malformed GraphicsCommand control block, invalid key character: 0x1b`, but after selecting a GIF file, it stops responding. Oh, and also file manager is stuck at using 1 CPU core until I kill the process.

Also see https://github.com/zellij-org/zellij/issues/3954.

**Comment by misaelaguayo** [comment]
I’ve noticed the placement issue. I think we might have to pass in some ansi codes to place it correctly. Also not sure why it doesn’t seem to clear. Might have to make some updates to the clear command as well. I’ll keep looking into it and and see if it can be in a useable state

**Comment by PartyWumpus** [comment]
Unicode placement support should be relatively trivial, but any other kind of positioning will need parsing, understanding and modifying of the request.

**Comment by valyntyler** [comment]
Can confirm, it works great on wezterm. Any chance we see the same kind of support for ghostty?

**Comment by ignamartinoli** [comment]
Tested on Kitty and Ghostty and t didn't work

**Comment by Raagh** [comment]
I really miss this feature, it sucks that I have to close zellij or open a regular tab in kitty everytime I want to view a markdown file with images.

**Comment by kungfusaini** [comment]
Bump on this ! It's literally the only thing preventing me switching from tmux!

**Comment by Raagh** [comment]
> Bump on this ! It's literally the only thing preventing me switching from tmux!

to me the other way around, is the thing that made me install tmux for the first time 😢

**Comment by misaelaguayo** [comment]
Could you guys checkout the changes on that [PR](https://github.com/misaelaguayo/zellij/pull/1) again? Tried doing a bit more work on it. 

Unfortunately, this does not look like it would be possible to do with a plugin. And there hasn't really been any response on if this could be implemented. I do not want to maintain a fork, but I think this work could be a good starting point

**Comment by mushrowan** [comment]
just flagging this as an issue that is pretty important for my day-to-day terminal use :)

**Comment by kofany** [comment]
keep we need this :P

**Comment by chrimon** [comment]
please add this

**Comment by devspaceship** [comment]
This is also kind of a deal breaker for me :/

**Comment by valyntyler** [comment]
> This is also kind of a deal breaker for me :/

This is one of the problems that determined my switch to ghostty's native multiplexing

**Comment by luccahuguet** [comment]
if you guys want, i can investigate forking zellij as a very thin wrapper upstream from the original project and using this fork by default in yazelix

https://github.com/luccahuguet/yazelix

I wont promise timelines or that I will get it done, only that I will investigate when other other more relevant issues in yazelix are taken care of, and it it could happen

then i could optionally try to merge this back to original zellij, and drop the fork when this gets merged

just open an issue in yazelix or upvote it if someone already created it

**Comment by Mach1212** [comment]
Sadge, guess I got to go to tmux or ghostty's native :(

<img width="1000" height="400" alt="Image" src="https://github.com/user-attachments/assets/f732efd8-5aba-4810-b0e1-0ef857812090" />

**Comment by Chikalona** [comment]
**Environment: Kitty + Zellij + Yazi (file manager)**

When running Yazi inside Zellij, image previews do not work because Yazi detects Zellij via the DA1 response instead of the underlying Kitty terminal. As a result, Yazi falls back to Chafa instead of using the Kitty Graphics Protocol (KGP).
Workaround: Starting Yazi with Zellij environment variables removed forces correct terminal detection:

`env -u ZELLIJ_SESSION_NAME -u ZELLIJ yazi`

A passthrough mode that transparently forwards terminal queries like DA1 to the underlying terminal would fix this — and benefit not just Yazi but any application that relies on terminal capability detection.

**Comment by almkx** [comment]
@Chikalona I wonder if there's a similar workaround for `kew`. This doesn't work (as expected):
```
env -u ZELLIJ_SESSION_NAME -u ZELLIJ yazi
```

**Comment by luccahuguet** [comment]
> if you guys want, i can investigate forking zellij as a very thin wrapper upstream from the original project and using this fork by default in yazelix
> 
> https://github.com/luccahuguet/yazelix
> 
> I wont promise timelines or that I will get it done, only that I will investigate when other other more relevant issues in yazelix are taken care of, and it it could happen
> 
> then i could optionally try to merge this back to original zellij, and drop the fork when this gets merged
> 
> just open an issue in yazelix or upvote it if someone already created it

yazelix has been daily driving this fork for several weeks now: https://github.com/luccahuguet/yazelix-zellij

it's been super smooth

there are two changes from original zellij:
- it implements the kitty graphics protocol in zellij, since ghostty is the yazelix default terminal and ghostty does not support sixel
- If YAZELIX_TERMINAL_WINDOW_TITLE_PREFIX is set, window titles becomes:
  - Yazelix - Ghostty (or other terminal emulator name) - <zellij-session-name>
  - Instead of how it is in upstream: <session> | <pane-title>
  - (basically harmless yazelix integration that should not bother other zellij non-yazelix users)

goal is to update it with upstream changes once a month

but there are other changes this fork will have  in the future

one change it will certainly have in the future is:
- when there is a stack of panes, creating or killing panes within it should not reorder them

basically because this bothers me quite a bit

so in total 2 changes, and soon to be 3 changes

I dont know if I will want to add more changes in future

and feel free to open any issues

**Comment by luccahuguet** [comment]
proof 

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/394fd015-fdf9-483b-b402-51521eb3dc36" />

**Comment by deadcoder0904** [comment]
Please add this feature by default. My use-case is using Pi with Ghostty and Zellij. See https://github.com/ghostty-org/ghostty/discussions/13362 for more.

And AI agents probably need to be able to see visually. I was using Codex and I love it because I can see the images that are being created or even just shown. If the project is image heavy. So that's why this is a must-have requirement now for me at least.

Also, this is the most upvoted reaction on this repo so everyone probably wants it.

**Comment by imroc** [comment]
I already switched to herdr, which has no such issue

**Comment by deadcoder0904** [comment]
@imroc What a coincidence. After I wrote this post & before your msg, I saw a couple of videos on Herdr + asked Chat/Claude about it as some X a/c recommended it yesterday only with 3 other tools (2 of which I knew but not Herdr) & I'm 100% switching too.

I think it does what Zellij does & more for my bare-minimum things. So long. Almost a year+ on Zellij.

**Comment by tesar-tech** [comment]
I made the switch, and this was one of the reasons. However, not everything is perfect in Herdr:

- It supports the Kitty graphics protocol. If you use Foot, or any other terminal emulator without Kitty support, images will not be displayed. I assume the same limitation would apply if Zellij supported the Kitty protocol.
- Floating panes are either missing or do not work as well as they do in Zellij. Zellij gets this right.
- Herdr closes tabs from right to left, while Zellij closes them based on the most recently used order. This behavior cannot be configured.
- Herdr does not support session restoration where you can press Enter to rerun the previous command.
- I have also encountered some issues with extra decorations in Neovim, such as squiggly lines under errors.
-  There is also no way how to move tabs left and right - at least not with keyboard - you can do it with your mouse.

Herdr has advantages as well, of course, but that is outside the scope of this discussion. I only wanted to share some issues that might be relevant if you are considering switching from Zellij to Herdr.

**Comment by airtonix** [comment]
> I already switched to herdr, which has no such issue

> I think it does what Zellij does & more for my bare-minimum things. So long. Almost a year+ on Zellij.

@deadcoder0904 @imroc 

It's a single purpose tool though, it doesn't do:

- stacked panes
- floating panes
- allow replacing the sidebar
- have a web sharing server

**Comment by deadcoder0904** [comment]
It doesn't matter guys, I've already switched & so far liking it.

I like the sidebar feature & less clutter on it due to Tabs so I can use 1 repo with 4 tabs & I know its from that repo only. No need for split tabs & all.

And my main goal is just to use 4-5 AI Agents. Nothing more than that as it becomes human bottleneck.

Plus it looks clean & has first-class support for AI Agents which is what I want. This video helped - https://www.youtube.com/watch?v=qnIu-Xu64H0 (and I follow only some of these tips like installing Pi extension, etc...)

Now let's keep this discussion focused otherwise it becomes Zellij vs Herdr lol.

**Comment by airtonix** [comment]
Too late you already put this thread in that box

**Comment by programmerlexi** [comment]
@airtonix @deadcoder0904 @tesar-tech @imroc 
Please stop this discussion or take it elsewhere and mark your comments regarding said discussion as off-topic.

**Comment by luccahuguet** [comment]
@programmerlexi is right. 

If any of you ladies and gentlemen want to continue this off-topic comparison, we can continue it here is this discussion I just created

https://github.com/zellij-org/zellij/discussions/5391

**Comment by Pshemas** [comment]
+1 on kitty image support / passthrough .

I've just hit this limitation when working on markdown files in the terminal - so I either have session restoration (thx to Zellij), or image preview (thx to Kitty Image Protocol support in the terminal)... but not both .

**Comment by imsnif** [comment]
Implemented: https://github.com/zellij-org/zellij/pull/5428

At the time of writing this is not yet released, I hope to cut a release soon.

**Comment by luccahuguet** [comment]
@imsnif is the best there is
